security/security_principles/README.md

```diff
@@ -0,0 +1,76 @@
+# Security Principles for Developers
+
+Security on the web relies on confidentiality, integrity, and availability. Considering each of these key pieces while we build helps us produce robust security control. The following security principles embody these ideals.
+
+## Defense in Depth
+
+**A secure application has multiple layers of security controls.**
+
+Defense in depth is the idea that security should have layers. Multiple layers of security controls make severe vulnerabilities difficult for an attacker to exploit.
+
+For example, a flawed administrative interface is unlikely to be vulnerable if it gates access to networks, checks for admin user authorization, and logs all access.
+
+## Positive Security Model
+
+**A secure application defines what is allowed and rejects everything else.**
+
+For example, if we’re using the positive security model while working with a form input, we should specify only what characters will be allowed. That may mean only alphanumeric characters are allowed or only numbers are allowed, for example.
+
+Another example is in user access control, where a user would be allowed access only to what is needed and denied access to everything else.
+
+## Fail Securely
+
+**A secure application handles errors and exceptions securely.**
+
+Handling errors securely is a key aspect of secure coding. Exceptions that occur while processing a security control deserve special attention. These exceptions must not allow behavior that the control wouldn’t normally allow. Any time we implement a security mechanism, there are three possible outcomes:
+
+Allow the operation
+Disallow the operation
+Exception
+
+Generally, we design security mechanisms so that failure, or exception, is the same as disallowing the operation.
+
+## Principle of Least Privilege
+
+**A secure application gives the least amount of privilege required.**
+
+This principle encompasses user rights and resource permissions.
+
+For example, when we add clients to Slack, we only give them access to one external client channel.
+
+Another example is a server that only needs access to a network, read access to a database table, and the ability to write a log. So those are the only permissions that the server should be granted.
+
+Keep in mind that it’s often easier to give additional permissions than to take away existing permissions.
+
+## Separation of Duties
+
+**A secure application separates duties, restricting the power of any individual.**
+
+Separation of duties manages conflict of interest by restricting any one individual’s power. It is best implemented by using role-based access control and by carefully defining what each role should be able to do.
+
+We can apply this method to job roles, where a developer develops, a QA tester tests, and a devops engineer deploys. This kind of separation ensures plenty of eyes on the code before it goes live.
```

Comment:
While the example applies on paper, the title-to-responsibility alignment is anathema to our principles. Everyone tests to create shared responsibility.

What if we used environments instead:

```suggestion
We can apply this method to environments, where a developers would not have administrative access to production environments. This kind of separation ensures production changes only occur through controlled means.
```

Reply:
TIL the word anathema. Great callout here.